Refactored gaming plugin for project consistency and rewrote coin flip function

leonthemisfit opened the pull request:
The plugin's docstrings were not consistent with the rest of the project and have
been updated to better reflect that. The coin flip function has also been rewritten
with new logic that estimates large flip amounts in a cleaner more maintainable
manner while still providing very similar results to the original method.

leonthemisfit committed:
Updated gaming plugin docstrings

Updates the docstrings in the gaming plugin to be more
consistent with the rest of the project.

leonthemisfit committed:
Refactored coin function to use string constants

The user facing strings were hardcoded into the function
rather than being constants. Having them as constants
improves readability and maintainability.

leonthemisfit committed:
Rewrote coin flip to use a cleaner method

The original way of calculating coin flips was more complex
and harder to read than needed. This changes it to use a
simpler and cleaner method that's still statistically close
enough to actual coin flip results.

leonthemisfit committed:
Removed clamp function from gaming plugin

The clamp function was originally necessary to deal with
some issues with the coin flip function. The new logic
for coin flip makes this no longer necessary and therefore
it can be removed.

leonthemisfit committed:
Added leonthemisfit to modified by

Added my name (leonthemisfit) to the modified by
comment because of making significant changes.

leonthemisfit committed:
Fixed comparison typo and added roll limit const to gaming plugin

There was a typo in the new coin flip logic where it was
supposed to run a proper simulation if flips were under
the roll limit but instead would only run if they were over.

Also moved hardcoded roll limit to a constant because it
is used in multiple places.

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -15,32 +16,30 @@
     """
     if n in ('f', 'F'):
-        return [random.randint(-1, 1) for _ in range(min(count, 100))]
+        return [random.randint(-1, 1) for _ in range(min(count, ROLL_LIMIT))]
```
Might want to change this to "fake" the result similar to the other checks, for consistency.

leonthemisfit replied:
I will look into finding a method to approximate the results. The method for other rolls will not work directly with the fudge dice because of the nature of the dice.

leonthemisfit replied:
This can be approximated using the same method as any other die with slight variations for the values associated with the fudge dice. These variations have been programmed into the function with a major refactor to adjust for them and keep the code base cleaner.

leonthemisfit committed:
Updated format strings to remove ' literals and added invalid roll const

As per request the format strings utilizing a single quote
within the string have been modified to remove the quotes
and use the {!r} formatter. The string for invalid rolls has
also been moved to a constant for consistency with the
rest of the plugin as well as improving readability.

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -55,25 +56,25 @@ def n_rolls(count, n):
     :type text: str
     """
-
     if hasattr(text, "groups"):
```
This check shouldn't be needed anymore as `dice` is never called with a regex match

leonthemisfit replied:
The current dice roll command is in the process of being rewritten in its entirety to provide cleaner code properly separated into individual functions so this change will no longer be necessary at that point, If need be this can be refactored first but I'd prefer to only refactor that code further if it's absolutely necessary since it's being replaced.

leonthemisfit committed:
Fixed incorrect roll count comparison

The n_rolls function was incorrectly checking if the number
of sides was less than the roll limit rather than the number
of rolls to perform.

leonthemisfit committed:
Added fudge approximation and refactored to match change

This adds the ability to approximate fudge rolls greater
than the roll limit in n_rolls. This also majorly refactors
n_rolls because the changes to fudge dice make it possible
to use the approximation calculations in a more generic
way that is applicable to both fudge and non fudge dice.

leonthemisfit committed:
Patches bug with bounds checking in .dice command

The wrong variable was being checked resulting in
simulating very large numbers rather than approximating
them. This patch fixes that.

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -15,38 +16,46 @@
+    fudge = n in ('f', 'F')
 
     if count < 100:
         return [random.randint(1, n) for _ in range(count)]
```
This will break if using a fudge dice rolled less than 100 times

leonthemisfit committed:
Fixed fudge dice issues with simulated rolls

Because 'F' is a letter and not a number if a person were to roll
a fudge die less than the roll limit the call to random.randint
would fail. This has been replaced with a correct checking
of fudge dice and ensuring the correct values are passed.

"Uh, Big Bird, F is not a number... F is a letter"
  --Kramit the Frog, Sensimilla Street

leonthemisfit committed:
Moved approximation explanation to a clearer location

The current location of the explanatory comment is actually
beneath a large chunk of where the approximation happens. This
is remedied by moving the comment above those approximations.

leonthemisfit committed:
Refactored n_rolls to have clearer param names

The original parameter names were unclear and has already
caused confusion over the process of cleaning this up, so a quick
refactor has been done to clear up confusion.

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -15,39 +16,53 @@
-    :type n: int | str
+
+    :type roll_cnt: int
+    :type sides: int | str
```
```suggestion
    :type sides: int | str
    :rtype: list[str]
```

leonthemisfit replied:
The return type of nrolls is a list of ints, not strings.

linuxdaemon replied:
Yes I realized just now that I mistyped that

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -138,24 +152,23 @@ def coin(text, notice, action):
+        if amount < ROLL_LIMIT:
+            heads = sum(random.randint(0, 1) for _ in range(amount))
+        else:
+            heads = int(amount * random.uniform(0.45, 0.55))
```
```suggestion
            heads = round(amount * random.uniform(0.45, 0.55))
```

linuxdaemon commented on plugins/gaming.py:
```diff
@@ -15,39 +16,53 @@
+
+    adj_var = (var * roll_cnt) ** 0.5
 
     return [int(random.normalvariate(mid, adj_var))]
```
```suggestion
    return [round(random.normalvariate(mid, adj_var))]
```

linuxdaemon reviewed (changes requested):
Using `round()` instead of `int()` is likely to produce more accurate results

leonthemisfit committed:
Changed calls to int to calls to round

The two places that int() was used to get the floor of a number
has been replaced with round because linuxdaemon.

leonthemisfit committed:
Added returned type to nrolls

This simply adds a return type in the docstring of the nrolls
function in gaming.py.

linuxdaemon reviewed (changes requested):
Please add some unit tests, something along the lines of https://gist.github.com/linuxdaemon/29cad9c3371f1e9182f3686d63a72170 should work.

leonthemisfit committed:
Moved approximation formulas to separate functions

In order to provide better documentation and readability as to
how rolls are approximated the formulas have been moved to
their own functions. This allows each formula to be better
documented in docstrings as well as giving them concise names
so that there's no reliance on just variable names to understand
how the values are being found.

leonthemisfit committed:
Moved approximation to separate function

In order to continue to improve readability and testability the
approximation of rolls has been moved to a separate function.
This allows n_rolls to be shorter and more readable.

leonthemisfit committed:
Moved dice roll simulation to its own function

The last step in this refactor is to move dice roll simulation to a
separate function. This final piece of refactoring represents all
logic being separated into logical units. This provides the best
means to document and test the various functions. This also
provides better maintainability.

leonthemisfit committed:
Refactored n_rolls to follow SESE

Rather than having multiple return statements which can often
cause confusion, the results of the two function calls are stored
and then returned at the end of the code block. This can also
help future changes where the returned values may be
manipulated for additional features or other unforeseen reasons.

leonthemisfit committed:
Rounded results of find_variance function

4 digits is plenty of precision for the calculations that use this to
be accurate. This is also done so that tests in docstrings will be
easier to read.

leonthemisfit committed:
Rounded the result of find_adjusted_variance()

Like find_variance  4 digits of precision is plenty for its use here
and it improves readability of test results.

leonthemisfit committed:
Added tests to utility function docstrings

This commit adds simple tests/examples to the docstrings of
util functions that are non-random (for simplicity) so that you
can get an idea of what kind of data they return.

leonthemisfit committed:
Updated tests so they correct pass in travis

There was a discrepancy between my local test results and the
test results from travis with the return type from find_midpoint.
The tests in the plugin have been updated to reflect the results
travis expects rather than my local machine.

linuxdaemon reviewed (changes requested):
Please expand test coverage according to https://codecov.io/gh/snoonetIRC/CloudBot/pull/391/src/plugins/gaming.py
